src/index.tsx

```diff
@@ -12,12 +12,23 @@ type Modules = ReturnType<qrcodegen.QrCode['getModules']>;
 type Excavation = {x: number; y: number; w: number; h: number};
 type ErrorCorrectionLevel = 'L' | 'M' | 'Q' | 'H';
 
-const ERROR_LEVEL_MAP: {[index: string]: qrcodegen.QrCode.Ecc} = {
+type ERROR_LEVEL_MAPPED_TYPE = {
+  [index in ErrorCorrectionLevel]: qrcodegen.QrCode.Ecc;
+};
+
+type GetImageSettingsReturnType =
+  | null
+  | (Excavation & {
+      excavation: Excavation | null;
+      opacity: number;
+    });
```

Comment:
I extracted the return type of the complex `getImageSettings` function and reused the already defined `Excavation`.

Reply:
Extracting the type is nice but…

This is not wrong per se, but does get really confusing to read. Really it's because I didn't name the `Excavation` type with this in mind. If you renamed it here to something like layout or coordinates, then the intersection would make sense when reading.